### ISSUE TITLE
Exceptionally large pages on split

### ISSUE THREAD

**Issue by wmbutler** [qpdf/qpdf#373]
Jay,

We have a case where the page split function is creating unusually large pages. This feels very similar to the issue where we had a common library or graphic getting replicated on each page.

See page 0037 where things get large. As before, happy to compensate you for any time or research you can put towards this. Also, I'm happy to provide any diagnostics. I'm unable to provide the source documents due to HIPAA constraints.

```
-rw-r--r--   1 bbutler  staff     19381 Oct  8 09:36 result--0001.pdf
-rw-r--r--   1 bbutler  staff     49322 Oct  8 09:36 result--0002.pdf
-rw-r--r--   1 bbutler  staff    102784 Oct  8 09:36 result--0003.pdf
-rw-r--r--   1 bbutler  staff    133750 Oct  8 09:36 result--0004.pdf
-rw-r--r--   1 bbutler  staff    128992 Oct  8 09:36 result--0005.pdf
-rw-r--r--   1 bbutler  staff    160950 Oct  8 09:36 result--0006.pdf
-rw-r--r--   1 bbutler  staff    246855 Oct  8 09:36 result--0007.pdf
-rw-r--r--   1 bbutler  staff    233045 Oct  8 09:36 result--0008.pdf
-rw-r--r--   1 bbutler  staff    138427 Oct  8 09:36 result--0009.pdf
-rw-r--r--   1 bbutler  staff    180721 Oct  8 09:36 result--0010.pdf
-rw-r--r--   1 bbutler  staff    174801 Oct  8 09:36 result--0011.pdf
-rw-r--r--   1 bbutler  staff    171179 Oct  8 09:36 result--0012.pdf
-rw-r--r--   1 bbutler  staff    177171 Oct  8 09:36 result--0013.pdf
-rw-r--r--   1 bbutler  staff    179457 Oct  8 09:36 result--0014.pdf
-rw-r--r--   1 bbutler  staff    191737 Oct  8 09:36 result--0015.pdf
-rw-r--r--   1 bbutler  staff    194079 Oct  8 09:36 result--0016.pdf
-rw-r--r--   1 bbutler  staff    189008 Oct  8 09:36 result--0017.pdf
-rw-r--r--   1 bbutler  staff    197976 Oct  8 09:36 result--0018.pdf
-rw-r--r--   1 bbutler  staff    200297 Oct  8 09:36 result--0019.pdf
-rw-r--r--   1 bbutler  staff    187104 Oct  8 09:36 result--0020.pdf
-rw-r--r--   1 bbutler  staff    189024 Oct  8 09:36 result--0021.pdf
-rw-r--r--   1 bbutler  staff    188693 Oct  8 09:36 result--0022.pdf
-rw-r--r--   1 bbutler  staff    190108 Oct  8 09:36 result--0023.pdf
-rw-r--r--   1 bbutler  staff    189985 Oct  8 09:36 result--0024.pdf
-rw-r--r--   1 bbutler  staff    189496 Oct  8 09:36 result--0025.pdf
-rw-r--r--   1 bbutler  staff    206060 Oct  8 09:36 result--0026.pdf
-rw-r--r--   1 bbutler  staff    187449 Oct  8 09:36 result--0027.pdf
-rw-r--r--   1 bbutler  staff    191961 Oct  8 09:36 result--0028.pdf
-rw-r--r--   1 bbutler  staff    191443 Oct  8 09:36 result--0029.pdf
-rw-r--r--   1 bbutler  staff    189064 Oct  8 09:36 result--0030.pdf
-rw-r--r--   1 bbutler  staff    184439 Oct  8 09:36 result--0031.pdf
-rw-r--r--   1 bbutler  staff    178311 Oct  8 09:36 result--0032.pdf
-rw-r--r--   1 bbutler  staff    182055 Oct  8 09:36 result--0033.pdf
-rw-r--r--   1 bbutler  staff    182643 Oct  8 09:36 result--0034.pdf
-rw-r--r--   1 bbutler  staff    144540 Oct  8 09:36 result--0035.pdf
-rw-r--r--@  1 bbutler  staff  46375154 Oct  8 09:36 result--0036.pdf
-rw-r--r--   1 bbutler  staff  46375689 Oct  8 09:37 result--0037.pdf
-rw-r--r--   1 bbutler  staff  46374980 Oct  8 09:37 result--0038.pdf
-rw-r--r--   1 bbutler  staff  46374496 Oct  8 09:38 result--0039.pdf
-rw-r--r--   1 bbutler  staff  46374426 Oct  8 09:38 result--0040.pdf
-rw-r--r--   1 bbutler  staff  44350641 Oct  8 09:38 result--0041.pdf
```

**Comment by jberkenbilt** [comment]
What might help would be to run `qpdf --json` on the whole file, then to run `qpdf input.pdf --pages . 1,3,4,36,37 -- output.pdf` and to run `qpdf --json` on output.pdf. This may give me a hint as to what presumably shared object is being replicated. If you inspect the json output and discover that it is too revealing, maybe we can work on figuring out a way to create sanitized json, such as by replacing all strings with some kind of hash or something. It would be great to be able to debug this kind of issue without having to have the source files since this type of thing does come up from time to time.

**Comment by wmbutler** [comment]
Will do. I'm pretty sure it's a single image that is being shared on multiple pages. I think the strategy would be to optimize the repeating image. Here is a partial dump from `pdfimages -list`.

I'll grab that json for you in a moment.

```
4736  4785 image    2556  3310  rgb     3   8  image  no       372  0   352   414  556K 2.2%
4737  4786 image     195    69  rgb     3   8  image  no       330  0    96    96 3880B 9.6%
4737  4787 image    2556  3310  rgb     3   8  image  no       373  0   352   414  573K 2.3%
4738  4788 image     195    69  rgb     3   8  image  no       330  0    96    96 3880B 9.6%
4738  4789 image    2552  3312  rgb     3   8  image  no       374  0   351   414  563K 2.3%
4739  4790 image     195    69  rgb     3   8  image  no       330  0    96    96 3880B 9.6%
4739  4791 image    2556  3312  rgb     3   8  image  no       376  0   352   414  582K 2.3%
4740  4792 image     195    69  rgb     3   8  image  no       330  0    96    96 3880B 9.6%
4740  4793 image    2556  3312  rgb     3   8  image  no       377  0   352   414  547K 2.2%
4741  4794 image     195    69  rgb     3   8  image  no       330  0    96    96 3880B 9.6%
4741  4795 image    2556  3308  rgb     3   8  image  no       378  0   352   414  574K 2.3%
4742  4796 image     195    69  rgb     3   8  image  no       330  0    96    96 3880B 9.6%
4742  4797 image    2556  3304  rgb     3   8  image  no       379  0   352   413  559K 2.3%
4743  4798 image     195    69  rgb     3   8  image  no       330  0    96    96 3880B 9.6%
4743  4799 image    2552  3304  rgb     3   8  image  no       380  0   351   413  570K 2.3%
4744  4800 image     195    69  rgb     3   8  image  no       330  0    96    96 3880B 9.6%
4744  4801 image    2556  3302  rgb     3   8  image  no       381  0   352   413  566K 2.3%
4745  4802 image     195    69  rgb     3   8  image  no       330  0    96    96 3880B 9.6%
4745  4803 image    2556  3312  rgb     3   8  image  no       382  0   352   414  577K 2.3%
4746  4804 image     195    69  rgb     3   8  image  no       330  0    96    96 3880B 9.6%
4746  4805 image    2556  3302  rgb     3   8  image  no       383  0   352   413  587K 2.4%
4747  4806 image     195    69  rgb     3   8  image  no       330  0    96    96 3880B 9.6%
4747  4807 image    2556  3310  rgb     3   8  image  no       384  0   352   414  578K 2.3%
4748  4808 image     195    69  rgb     3   8  image  no       330  0    96    96 3880B 9.6%
4748  4809 image    2556  3312  rgb     3   8  image  no       385  0   352   414  562K 2.3%
4749  4810 image     195    69  rgb     3   8  image  no       330  0    96    96 3880B 9.6%
4749  4811 image    2556  3304  rgb     3   8  image  no       387  0   352   413  570K 2.3%
4750  4812 image     195    69  rgb     3   8  image  no       330  0    96    96 3880B 9.6%
4750  4813 image    2556  3312  rgb     3   8  image  no       388  0   352   414  538K 2.2%
4751  4814 image     195    69  rgb     3   8  image  no       330  0    96    96 3880B 9.6%
4751  4815 image    2556  3304  rgb     3   8  image  no       389  0   352   413  562K 2.3%
4752  4816 image     195    69  rgb     3   8  image  no       330  0    96    96 3880B 9.6%
4752  4817 image    2552  3304  rgb     3   8  image  no       390  0   351   413  568K 2.3%
4753  4818 image     195    69  rgb     3   8  image  no       330  0    96    96 3880B 9.6%
4753  4819 image    2560  3296  rgb     3   8  image  no       391  0   353   412  577K 2.3%
4754  4820 image     195    69  rgb     3   8  image  no       330  0    96    96 3880B 9.6%
4754  4821 image    2556  3304  rgb     3   8  image  no       392  0   352   413  570K 2.3%
4755  4822 image     195    69  rgb     3   8  image  no       330  0    96    96 3880B 9.6%
4755  4823 image    2556  3308  rgb     3   8  image  no       393  0   352   414  221K 0.9%
4756  4824 image     195    69  rgb     3   8  image  no       330  0    96    96 3880B 9.6%
4756  4825 image    2556  3312  rgb     3   8  image  no       394  0   352   414  550K 2.2%
4757  4826 image     195    69  rgb     3   8  image  no       330  0    96    96 3880B 9.6%
4757  4827 image    4032  3024  rgb     3   8  image  no       339  0   555   554 5791K  16%
4758  4828 image     195    69  rgb     3   8  image  no       330  0    96    96 3880B 9.6%
4758  4829 image    2548  3304  rgb     3   8  image  no       395  0   351   413  563K 2.3%
4759  4830 image     195    69  rgb     3   8  image  no       330  0    96    96 3880B 9.6%
4759  4831 image    1702  2207  rgb     3   8  image  no       396  0   235   276  317K 2.9%
4760  4832 image     195    69  rgb     3   8  image  no       330  0    96    96 3880B 9.6%
4760  4833 image    2556  3302  rgb     3   8  image  no       398  0   352   413  560K 2.3%
4761  4834 image     195    69  rgb     3   8  image  no       330  0    96    96 3880B 9.6%
4761  4835 image    2552  3312  rgb     3   8  image  no       399  0   351   414  170K 0.7%
4762  4836 image     195    69  rgb     3   8  image  no       330  0    96    96 3880B 9.6%
4762  4837 image    2552  3312  rgb     3   8  image  no       400  0   351   414  421K 1.7%
4763  4838 image     195    69  rgb     3   8  image  no       330  0    96    96 3880B 9.6%
4764  4839 image     195    69  rgb     3   8  image  no       330  0    96    96 3880B 9.6%
4764  4840 image    2550  3300  rgb     3   8  image  no       401  0   391   391  593K 2.4%
4765  4841 image     195    69  rgb     3   8  image  no       330  0    96    96 3880B 9.6%
4765  4842 image    2550  3300  rgb     3   8  image  no       402  0   391   391  419K 1.7%
4766  4843 image     195    69  rgb     3   8  image  no       330  0    96    96 3880B 9.6%
4766  4844 image    2550  3300  rgb     3   8  image  no       403  0   391   391  292K 1.2%
4767  4845 image     195    69  rgb     3   8  image  no       330  0    96    96 3880B 9.6%
4767  4846 image    2550  3300  rgb     3   8  image  no       404  0   391   391  480K 1.9%
4768  4847 image     195    69  rgb     3   8  image  no       330  0    96    96 3880B 9.6%
4768  4848 image    2550  3300  rgb     3   8  image  no       405  0   391   391  511K 2.1%
4769  4849 image     195    69  rgb     3   8  image  no       330  0    96    96 3880B 9.6%
4769  4850 image    2550  3300  rgb     3   8  image  no       406  0   391   391  495K 2.0%
4770  4851 image     195    69  rgb     3   8  image  no       330  0    96    96 3880B 9.6%
4770  4852 image    2550  3300  rgb     3   8  image  no       407  0   391   391 24.0K 0.1%
4771  4853 image     195    69  rgb     3   8  image  no       330  0    96    96 3880B 9.6%
```

**Comment by jberkenbilt** [comment]
Does `--optimize-images` help? Another option to consider would be adding the ability for qpdf to downsample the image. I created an image manipulation library around the same time I originally created qpdf. There are, of course, lots of good image libraries out there. My library had a small handful of features in comparison to the "real" ones, but I wrote it specifically to work with pipelines and uses the exact same API as qpdf's Pipeline classes. Adding a simple downsampler to qpdf would be possible, even if it's a little out of sync with its usual feature set. Shrinking the size of PDF files is an often requested feature. I had written code for one customer once that generated PDFs with all the images replaced with 75 dpi images for use in areas without high bandwidth Internet access. Anyway, just a thought.

Are you using qpdf 9.0.x yet? I remember that you hadn't taken some upgrades because of performance regressions, but I think these are all fixed now.

**Comment by wmbutler** [comment]
We are still using 8.4.2. I will upgrade to 9.0 and test it out.

`--optimize-images` does not improve things. The result file ends up being close to the same size.

I can get the file down to 22MB from 80MB by running a `gs` command taking chopping down the image resolution to screen. This of course would work fine for documents where the text was native with images embedded separately. This strategy would NOT work well for cases where the entire page was a scanned image because at screen resolution, I'm afraid OCR would suffer heavily.

The resulting json file from the command you offered above was 45MB for the entire file. The smaller json file for pages 1,3,4,36,37 is attached here.


[output_small.txt](https://github.com/qpdf/qpdf/files/3704122/output_small.txt)

**Comment by jberkenbilt** [comment]
In the job where I developed qpdf and the image library, one of the main things was to take scanned images and prepare them for OCR. I had created an algorithm that created an image histogram to figure out a threshold, used bicubic interpolation to upsample the image by a factor of two, and the convert color -> grayscale -> bitonal and finally to compress with CCITT G4. The result was typically a page image that was on the order of 100K, sometimes with OCR quality that exceeded what you'd get on the original grayscale. This was 17 years ago, so the technology has changed a lot. Building all that into qpdf would be a bit much, but maybe something like that could fit into your workflow somewhere.
